formats/log/note.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"fmt"
+
+	"golang.org/x/mod/sumdb/note"
+)
+
+// CheckpointParser is responsible for parsing checkpoints for a specific log.
+// Any valid note must have been signed by all verifiers known by this parser.
+type CheckpointParser struct {
+	logVerifier       note.Verifier
+	logID             string
+	otherVerifiers    note.Verifiers
+	requiredOtherSigs int
+}
+
+// NewCheckpointParser creates a parser for checkpoints that must be signed by the log
+// which is verified by logVKey, and *all* of the other verifiers under otherVerifiers.
+func NewCheckpointParser(logVKey string, logID string, otherVerifiers ...string) (CheckpointParser, error) {
+	logVerifier, err := note.NewVerifier(string(logVKey))
+	if err != nil {
+		return CheckpointParser{}, fmt.Errorf("NewVerifier(%s): %v", logVKey, err)
+	}
+
+	ovs := make([]note.Verifier, len(otherVerifiers))
+	for i, k := range otherVerifiers {
+		ovs[i], err = note.NewVerifier(k)
+		if err != nil {
+			return CheckpointParser{}, fmt.Errorf("NewVerifier(%s): %v", k, err)
+		}
+	}
+	return CheckpointParser{
+		logVerifier:       logVerifier,
+		logID:             logID,
+		otherVerifiers:    note.VerifierList(ovs...),
+		requiredOtherSigs: len(otherVerifiers),
+	}, nil
+}
+
+// Parse returns a checkpoint parsed from the raw note, and any error if the note could
+// not be parsed, or any of the required signatures were missing.
+func (p CheckpointParser) Parse(chkptRaw []byte) (Checkpoint, *note.Note, error) {
+	r := &Checkpoint{}
+	n, err := note.Open(chkptRaw, note.VerifierList(p.logVerifier))
+	if err != nil {
+		return *r, n, fmt.Errorf("failed to verify log signature on checkpoint: %v", err)
+	}
+	if _, err := r.Unmarshal([]byte(n.Text)); err != nil {
+		return *r, n, fmt.Errorf("failed to unmarshal new checkpoint: %v", err)
+	}
+	// TODO(mhutchinson): This assumes the first line is the LogID.
+	// This is proposal 2B. If we put the LogID further in the message (e.g. 4th line; 1A)
+	// then this should be updated.
+	if r.Ecosystem != p.logID {
+		return *r, n, fmt.Errorf("got logID %q but expected %q", r.Ecosystem, p.logID)
+	}
+
+	if p.requiredOtherSigs == 0 {
+		return *r, n, nil
+	}
+
+	if n, err = note.Open(chkptRaw, p.otherVerifiers); err != nil {
```

Comment:
I can see the reasoning behind this but it seems awkward to have two calls to `note.Open`, especially when the second one comes after you've already parsed the note.  If you did one call to `note.Open` with both the log and other verifiers then you could also remove lines 73-75, but on the other hand I guess you'd have no way to distinguish between a log or other signature failing which would make the error messaging less helpful.  So maybe I'm just saying I wish Note made it easier to find out which signatures verified and there is no actual suggestion here :).

Reply:
I realize it looks more awkward in here, but if it puts our minds at rest, the performance is pretty identical (if anything, this way seems faster, which is odd):

Benchmark with only one `note.Open` (https://github.com/mhutchinson/trillian-examples/pull/new/checkpointParserSimpler):
```
BenchmarkParse/bare_minimum-4                 7752        174560 ns/op        1024 B/op          14 allocs/op
BenchmarkParse/one_verifier-4                 4048        395803 ns/op        1520 B/op          18 allocs/op
BenchmarkParse/many_verifiers-4               1612        674900 ns/op        2450 B/op          25 allocs/op
BenchmarkParse/many_sigs,_1_verifier-4      4126        347926 ns/op        3953 B/op          41 allocs/op
```

Benchmark from this code with two `note.Open`:
```
BenchmarkParse/bare_minimum-4                 8337        144167 ns/op        1504 B/op          17 allocs/op
BenchmarkParse/one_verifier-4                 3813        304662 ns/op        3056 B/op          35 allocs/op
BenchmarkParse/many_verifiers-4               2007        578182 ns/op        5154 B/op          55 allocs/op
BenchmarkParse/many_sigs,_1_verifier-4      4092        298907 ns/op        8017 B/op          81 allocs/op
```

Reply:
I think there's a slight functional difference with the current impl in that the returned `n` will appear not to have had a valid sig from the log in its `Sigs` array (since we're overwriting the `n` which did have that record).

It's implicit that it did, of course, but maybe a bit weird that it's not present?

Reply:
I've changed this behaviour now. There's only one call to note.Open and all matching signatures are returned.